### ISSUE TITLE
Inference time vs torch w/regard to batch_size and BatchNorm

### ISSUE THREAD

**Issue by jogepari** [microsoft/onnxruntime#12130]
**Describe the bug**
Noticed that my simple tabular model having BatchNorm layers in it performs slower via ORT than Torch with increasing batch size (even though being significantly faster when batch size is small). Apparently, BatchNorm slows down ORT model way more severely than Torch with batch size >10. What could be the cause of it?

Exporting from torch based on [https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html](url)

Measurements might not be too precise, but probably enough to demonstrate the point.

```
Inference time (per batch):

                torch                onnx          
               no_BN        BN     no_BN        BN
batch_size                                        
1           0.000294  0.000371  0.000029  0.000043
10          0.000353  0.000560  0.000055  0.000159
100         0.000659  0.000870  0.000216  0.000991
1000        0.002460  0.002643  0.001109  0.009028
10000       0.032228  0.037158  0.010602  0.083429
100000      0.270563  0.329899  0.135601  0.880045
1000000     2.698562  3.202170  1.544639  7.719083
```

![onnx_bn_slowdown](https://user-images.githubusercontent.com/72675798/178014699-c48e2dd4-51ff-48cc-b9c6-efea16571b08.png)

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 11 (same on Linux though)
- ONNX Runtime installed from (source or binary): pip, onnxruntime==1.11.1
- Python version: 3.9.12 (same on 3.7.12)

**To Reproduce**
<details>
  <summary>Python code</summary>
  
```python
import pandas as pd
import numpy as np

import torch
from torch import nn
import onnx
import onnxruntime

from matplotlib import pyplot as plt
import seaborn as sns
import timeit
from functools import partial
import warnings

plt.style.use('seaborn')

class SimpleBlock(nn.Module):
    def __init__(self, n_out, batchnorm=True, dropout_p=0.1):
        super().__init__()
        layers_list = [
            nn.LazyLinear(n_out),
            nn.ReLU(),
            nn.Dropout(dropout_p)
        ]
        if batchnorm:
            layers_list.insert(0, nn.LazyBatchNorm1d())
        
        self.layers = nn.Sequential(*layers_list)
    
    def forward(self, x):
        return self.layers.forward(x)
        

class SimpleNet(nn.Module):
    def __init__(self, n_out, n_hidden=100, layers_hidden=10, batchnorm=True, dropout_p=0.1):
        super().__init__()
        layers_list = [nn.LazyLinear(n_hidden)] 
        layers_list.extend([SimpleBlock(n_hidden, batchnorm, dropout_p) for _ in range(layers_hidden)])
        layers_list.append(nn.LazyLinear(n_out))
        self.layers = nn.Sequential(*layers_list)
    
    def forward(self, x):
        return self.layers.forward(x)
    

def get_models(X_export, **net_kwargs):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        torch_model = SimpleNet(**net_kwargs)
    torch_model.eval()
    torch_out = torch_model(X_export)
    
    torch.onnx.export(torch_model,               # model being run
                  X_export,                         # model input (or a tuple for multiple inputs)
                  "SimpleNet.onnx",   # where to save the model (can be a file or file-like object)
                  export_params=True,        # store the trained parameter weights inside the model file
                  opset_version=15,          # the ONNX version to export the model to
                  do_constant_folding=True,  # whether to execute constant folding for optimization
                  input_names = ['input'],   # the model's input names
                  output_names = ['output'], # the model's output names
                  dynamic_axes={'input' : {0 : 'batch_size'},    # variable length axes
                                'output' : {0 : 'batch_size'}},)
    
    onnx_model = onnx.load("SimpleNet.onnx")
    onnx.checker.check_model(onnx_model)
    
    ort_session = onnxruntime.InferenceSession("SimpleNet.onnx")
    ort_input_name = ort_session.get_inputs()[0].name

    def to_numpy(tensor):
        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()
    
    # compute ONNX Runtime output prediction
    ort_input_name = ort_session.get_inputs()[0].name
    ort_inputs = {ort_input_name: to_numpy(X_export)}
    ort_outs = ort_session.run(None, ort_inputs)

    # compare ONNX Runtime and PyTorch results
    np.testing.assert_allclose(to_numpy(torch_out), ort_outs[0], rtol=1e-03, atol=1e-05)

    print("Exported model has been tested with ONNXRuntime, and the result looks good!")
    
    return torch_model, ort_session


def onnx_predict(data, ort_session):
    ort_input_name = ort_session.get_inputs()[0].name
    ort_inputs = {ort_input_name: data}
    ort_outs = ort_session.run(None, ort_inputs)
    return ort_outs[0]


def torch_predict(data, torch_model):
    with torch.no_grad():
        return torch_model.forward(data)
    
    
def main():
    in_dim = 100
    max_batch_exp = 6
    in_samples = 10 ** max_batch_exp
    X = torch.rand((in_samples, in_dim))
    X_export = X[:1000]
    
    torch_model, onnx_model = get_models(X_export, batchnorm=False, n_out=1)
    torch_model_bn, onnx_model_bn = get_models(X_export, batchnorm=True, n_out=1)
    
    inference_funcs = {
    'torch': [partial(torch_predict, torch_model=torch_model), 
              partial(torch_predict, torch_model=torch_model_bn)],
    'onnx': [partial(onnx_predict, ort_session=onnx_model), 
             partial(onnx_predict, ort_session=onnx_model_bn)]
    }
    bn_str = {False: 'no_BN', True: "BN"}
    
    batch_sizes = np.logspace(0, max_batch_exp, num=max_batch_exp+1, dtype=np.int32)
    results = pd.DataFrame(
        data = np.zeros((max_batch_exp + 1, 4)), 
        index = pd.Index(batch_sizes, name='batch_size'),
        columns=pd.MultiIndex.from_product([['torch', 'onnx'], ['no_BN', 'BN'],]))
    
    for batch_size in batch_sizes:
        print(f'Inference with batch size {batch_size}...')
        batch = X[:batch_size]
        number = min(1000, 10 ** 6 // batch_size)
    
        for model in('torch', 'onnx'):
            if model == 'onnx':
                batch = batch.numpy()
            for batchnorm in (False, True):
                t = timeit.timeit('inference_funcs[model][batchnorm](batch)', 
                                  globals=locals(), 
                                  number=number)
                results.loc[batch_size, (model, bn_str[batchnorm])] = t / number
    
    print('\nInference time (per batch):\n\n', results)
    
    speed_vs_torch = np.log(results['onnx'] / results['torch'])
    speed_vs_torch_melt = pd.melt(speed_vs_torch.reset_index(),
                      id_vars='batch_size', var_name='BN', 
                      value_name='log_time_onnx_vs_torch')
    
    bn_slowdown = results.loc[:, (slice(None), 'BN')].droplevel(1, 1) / results.loc[:, (slice(None), 'no_BN')].droplevel(1, 1)
    bn_slowdown_melt = pd.melt(bn_slowdown.reset_index(), id_vars='batch_size', var_name='model', value_name='bn_slowdown')
    
    fig, axs = plt.subplots(1, 2, figsize=(16, 6))
    sns.barplot(data=speed_vs_torch_melt, x='batch_size', y='log_time_onnx_vs_torch',
                hue='BN', ax=axs[0]).set_title('Inference: log(onnx_time / torch_time)')
    sns.pointplot(data=bn_slowdown_melt, x='batch_size', y='bn_slowdown', 
                  hue='model', ax=axs[1]).set_title('BN slowdown');
    

if __name__ == '__main__':
    main()
```
</details>

**Expected behavior**
Being faster than Torch regardless of batch size and BatchNorm.

**Comment by jogepari** [comment]
@tianleiwu, thank you for the reply! But I'm measuring inference times on CPU , not sure what CUDA has to do with it.

**Comment by fxmarty** [comment]
@jogepari Did you find a workaround to this?

**Comment by jogepari** [comment]
> @jogepari Did you find a workaround to this?

Unfortunately, no. 

Luckily, in production conditions there should be no large batches probably, so ORT is still doing well, but when running tests - this slowdown is quite noticeable.
